pilot/pkg/networking/apigen/apigen.go

```diff
@@ -0,0 +1,162 @@
+// default, using the same mechanism that EDS is using, i.e. sending only changed resources
+// in a push.
+type ApiGenerator struct {
+}
+
+// TODO: take 'updates' into account, don't send pushes for resources that haven't changed
+// TODO: support WorkloadEntry - to generate endpoints (equivalent with EDS)
+
+// Generate implements the generate method for high level APIs, like Istio config types.
+// This provides similar functionality with MCP and :8080/debug/configz.
+//
+// Names are based on the current resource naming in istiod.
+// IMPORTANT: based on lessons from MCP, the response has as last element a 'sync' resource.
+// Currently a 'sync' just an empty one.
+// This will allow avoiding extremely large packet sizes, and chunking. Not yet reflected in
+// the XDS implementation, but useful to get into the initial protocol. The behavior is specific
+// to this type of resources - and not prohibited by the ADS protocol, EDS has a similar behavior.
+func (g *ApiGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
+	resp := []*any.Any{}
+
+	// Example: networking.istio.io/v1alpha3/VirtualService
+	// Note: this is the style used by MCP and its config. Pilot is using 'Group/Version/Kind' as the
+	// key, which is similar.
+	//
+	// The actual type in the Any should be a real proto - which is based on the generated package name.
+	// For example: type is for Any is 'type.googlepis.com/istio.networking.v1alpha3.EnvoyFilter
+	// We use: networking.istio.io/v1alpha3/EnvoyFilter
+	gvk := strings.SplitN(w.TypeUrl, "/", 3)
+	if len(gvk) == 3 {
+		rgvk := resource.GroupVersionKind{
+			Group:   gvk[0],
+			Version: gvk[1],
+			Kind:    gvk[2],
+		}
+		if w.TypeUrl == collections.IstioMeshV1Alpha1MeshConfig.Resource().GroupVersionKind().String() {
+			meshAny, err := types.MarshalAny(push.Mesh)
+			if err == nil {
+				resp = append(resp,&any.Any{
+					TypeUrl: w.TypeUrl,
+					Value:  meshAny.Value,
+				})
+			}
+			return resp
+		}
+
+		cfg, err := push.IstioConfigStore.List(rgvk, "")
+		if err != nil {
+			log.Warnf("ADS: Unknown watched resources %s %v", w.TypeUrl, err)
+			return resp
+		}
+		for _, c := range cfg {
+			// Right now model.Config is not a proto - until we change it, mcp.Resource.
+			// This also helps migrating MCP users.
+
+			b, err := configToResource(&c)
+			if err != nil {
+				log.Warna("Resource error ", err, " ", c.Namespace, "/", c.Name)
+				continue
+			}
+			bany, err := types.MarshalAny(b)
+			if err == nil {
+				resp = append(resp, &any.Any{
+					TypeUrl: bany.TypeUrl,
+					Value:  bany.Value,
+				})
+			} else {
+				log.Warna("Any ", err)
+			}
+		}
+
+		// TODO: MeshConfig, current dynamic ProxyConfig (for this proxy), Networks
+
+		if w.TypeUrl == collections.IstioNetworkingV1Alpha3Serviceentries.Resource().GroupVersionKind().String() {
```

Comment:
The use case is the same as MCP - we want a way to sync Istio resources. The debug interface is not really the way to do it. 

The 'main' PR is using this to move the LDS/RDS/CDS generation to agent - not clear if that will get merged any time soon, but it's important to get data on how it works. 

All use cases for MCP remain - for example with this PR the agent is able to get the MeshConfig at startup, instead of having it injected. It could also get a ProxyConfig, or the bootstrap template - as we discussed in the past. 

The other use case I'm interested in is to pull the ServiceEntries and resolve DNS in the agent.

Reply:
I really don't feel like we should merge a new API which we may be stuck with for a long time without further discussion. I think I am the only one that has looked at this PR and I certainly don't want to be the only person that looked at this

Reply:
Moved the apigen to separate PR